contracts/zero-ex/contracts/src/features/MetaTransactionsFeatureV2.sol

```diff
@@ -0,0 +1,620 @@
+        for (uint256 i = 0; i < state.mtx.fees.length; ++i) {
+            _transferERC20TokensFrom(state.mtx.feeToken, state.mtx.signer, state.mtx.fees[i].recipient, state.mtx.fees[i].amount);
+        }
+
+        // Execute the call based on the selector.
+        state.selector = state.mtx.callData.readBytes4(0);
+        if (state.selector == ITransformERC20Feature.transformERC20.selector) {
+            returnResult = _executeTransformERC20Call(state);
+        } else if (state.selector == INativeOrdersFeature.fillLimitOrder.selector) {
+            returnResult = _executeFillLimitOrderCall(state);
+        } else if (state.selector == INativeOrdersFeature.fillRfqOrder.selector) {
+            returnResult = _executeFillRfqOrderCall(state);
+        } else if (state.selector == IMultiplexFeature.multiplexBatchSellTokenForToken.selector) {
+            returnResult = _executeMultiplexBatchSellTokenForTokenCall(state);
+        } else if (state.selector == IMultiplexFeature.multiplexBatchSellTokenForEth.selector) {
+            returnResult = _executeMultiplexBatchSellTokenForEthCall(state);
+        } else if (state.selector == IMultiplexFeature.multiplexMultiHopSellTokenForToken.selector) {
+            returnResult = _executeMultiplexMultiHopSellTokenForTokenCall(state);
+        } else if (state.selector == IMultiplexFeature.multiplexMultiHopSellTokenForEth.selector) {
+            returnResult = _executeMultiplexMultiHopSellTokenForEthCall(state);
+        } else {
+            LibMetaTransactionsRichErrors.MetaTransactionUnsupportedFunctionError(state.hash, state.selector).rrevert();
+        }
+        emit MetaTransactionExecuted(state.hash, state.selector, state.mtx.signer, state.mtx.sender);
+    }
+
+    /// @dev Validate that a meta-transaction is executable.
+    function _validateMetaTransaction(ExecuteState memory state) private view {
+        // Must be from the required sender, if set.
+        if (state.mtx.sender != address(0) && state.mtx.sender != state.sender) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionWrongSenderError(state.hash, state.sender, state.mtx.sender)
+                .rrevert();
+        }
+        // Must not be expired.
+        if (state.mtx.expirationTimeSeconds <= block.timestamp) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionExpiredError(state.hash, block.timestamp, state.mtx.expirationTimeSeconds)
+                .rrevert();
+        }
+
+        if (LibSignature.getSignerOfHash(state.hash, state.signature) != state.mtx.signer) {
+            LibSignatureRichErrors
+                .SignatureValidationError(
+                    LibSignatureRichErrors.SignatureValidationErrorCodes.WRONG_SIGNER,
+                    state.hash,
+                    state.mtx.signer,
+                    // TODO: Remove this field from SignatureValidationError
+                    //       when rich reverts are part of the protocol repo.

```

Comment:
Open TODO

Reply:
This is sadly copy/pasted from the original metatransaction v1. Happy to open a TODO if this is something we actually want to fix.